transcom/mymove
# MB-16179 fix internal api upload amended orders authz issue

## [Jira ticket](https://dp3.atlassian.net/browse/MB-16179)

## Summary

We have a [discovery doc that outlines the API endpoints that have gaps in authz checks, as well as a plan to address them](https://dp3.atlassian.net/wiki/spaces/MT/pages/2115043329/Discovery+-+Customer+Internal+API+Authorization+Issues) 🔒. Part of what we decided was that we would include user information as part of the query when possible in order to let the DB handle the search and avoid us even loading the data (vs loading the data and then checking if the service member matches). This also easily sets us up to return a 404 when the user doesn't match, which is good because this way we won't even confirm that the ID they tried using is associated with anything (in this case we use order ID).

We didn't have a good example of what this should fully look like, but I'd put together a [post in slack that pieced together bits of code as examples](https://ustcdp3.slack.com/archives/C0509T4E8AU/p1685574194523799) 🔒.

<details>
<summary>Main Changes (i.e. things that explicitly satisfy AC)</summary>

* [fix: return error if user tries to upload amended orders for another sm](https://github.com/transcom/mymove/pull/10919/commits/249b60eb7d784165f66fd3c7e0d678339376a52e)
    * Previously we didn't check if the service member making the request matched the service member on the orders. We now check if the request is a customer app request (since we're doing this at a service object level), and if so, include their service member ID as part of the query. We'll return a 404 if it doesn't match, which is better security since we won't confirm for other users that they even have a valid order ID.
* [test (integration): add integration tests for amended orders uploads with routing & middleware enabled](https://github.com/transcom/mymove/pull/10919/files#diff-8365e1efe886d43016d781c96709927e8feb091914839e26123525fc2444a58e)
    * Linking to file diff because this was done across a couple of commits.

</details>

<details>
<summary>Other Changes</summary>

* There are changes in this PR that make it easier to make insecure method requests (e.g. POST, PUT, etc.) in the integration tests that have routing & middleware enabled. These changes came in via #10975 and are labeled as such in PR comments.
    * Discussion thread that led to this: https://github.com/transcom/mymove/pull/10919#discussion_r1244585956
* [tests (handlers): add more testing for amended orders upload](https://github.com/transcom/mymove/pull/10919/commits/5d4897864a00452e290785a8d252eb2c02a69ccc)
    * Helps us be sure the other changes don't break things that are currently working. Previously only had one success test case.
* [fix (handlers): fix improperly raised error in amended orders upload](https://github.com/transcom/mymove/pull/10919/commits/94fc4129a03cda7ba082f66502d5294f0bd5861f)
    * We were using an error function that expected to be passed an error, but we were passing in `nil`, which raised a different server error than we would have originally. Since that handler error function isn't what we typically do anymore, I just skipped calling it and instead now return a server error directly.
* [feat (test): add s3 file storer for routing integration tests](https://github.com/transcom/mymove/pull/10919/commits/81914ac8832aa5fe90825f7bf537e7c25b67c950)
    * Discussion thread: https://github.com/transcom/mymove/pull/10919#discussion_r1244585436

</details>

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Start server and customer client.
2. Log in as a service member that has just submitted their move, e.g. one of the ones that has the `complete@ppm.submitted` email.
3. Look at the `logged_in` response info to find their orders ID. The tree of data should be service_member.orders[0].id
    1.  
![image](https://github.com/transcom/mymove/assets/35938642/c923f404-6c6f-48b9-8f36-3da316ce7883)
4. Log out and log in as another service member to get their cookie info for postman. Load into postman based on however you set up postman.
    1. [Docs for setting up postman](https://transcom.github.io/mymove-docs/docs/tools/postman/setting-up-postman)
5. Make a patch request in postman to `/orders/{ordersId}/upload_amended_orders` 
    1. [Docs for sending files via postman](https://www.postman.com/postman/workspace/postman-answers/documentation/13455110-00378d5c-5b08-4813-98da-bc47a2e6021d)
    2. Since this is going to be a patch request, you'll need to [set the `X-CSRF-TOKEN` header to be the `masked_gorilla_csrf` cookie value](https://transcom.github.io/mymove-docs/docs/tools/postman/intercepting-cookies-for-postman#form-a-working-get-and-patch-request-in-postman).
6. Verify you get a 404 response.
    1. 
![image](https://github.com/transcom/mymove/assets/35938642/1587afe8-eea1-4187-901b-1640626dc06a)
    2. If you try the above steps on the main branch, you'll get a 201 instead indicating the amended orders were uploaded successfully.
 

### Backend
